signature/jws/jwt.go

```diff
@@ -0,0 +1,152 @@
+package jws
+
+import (
+	"crypto"
+	"crypto/x509"
+	"encoding/base64"
+	"errors"
+	"fmt"
+
+	"github.com/golang-jwt/jwt/v4"
+	"github.com/notaryproject/notation-core-go/signature"
+)
+
+// signingMethod is the interface for jwt.SigingMethod with additional method to
+// access certificate chain after calling Sign().
+type signingMethod interface {
+	jwt.SigningMethod
+
+	// CertificateChain returns the certificate chain.
+	//
+	// It should be called after calling Sign().
+	CertificateChain() ([]*x509.Certificate, error)
+
+	// PrivateKey returns the private key.
+	PrivateKey() crypto.PrivateKey
+}
+
+// remoteSigningMethod wraps the remote signer to be a SigningMethod
+type remoteSigningMethod struct {
+	signer    signature.Signer
+	certs     []*x509.Certificate
+	algorithm string
+}
+
+func newRemoteSigningMethod(signer signature.Signer) (signingMethod, error) {
+	algorithm, err := extractJwtAlgorithm(signer)
+	if err != nil {
+		return nil, err
+	}
+	return &remoteSigningMethod{
+		signer:    signer,
+		algorithm: algorithm,
+	}, nil
+}
+
+// Verify doesn't need to be implemented.
+func (s *remoteSigningMethod) Verify(signingString, signature string, key interface{}) error {
+	return errors.New("not implemented")
+}
+
+// Sign hashes the signingString and call the remote signer to sign the digest.
+func (s *remoteSigningMethod) Sign(signingString string, key interface{}) (string, error) {
+	// sign by external signer
+	sig, certs, err := s.signer.Sign([]byte(signingString))
+	if err != nil {
+		return "", err
+	}
+	s.certs = certs
+	return base64.RawURLEncoding.EncodeToString(sig), nil
+}
+
+// Alg implements jwt.SigningMethod interface.
+func (s *remoteSigningMethod) Alg() string {
+	return s.algorithm
+}
+
+// CertificateChain returns the certificate chain.
+//
+// It should be called after Sign().
+func (s *remoteSigningMethod) CertificateChain() ([]*x509.Certificate, error) {
+	if s.certs == nil {
+		return nil, &signature.InvalidSignRequestError{Msg: "certificate chain is not set"}
+	}
+	return s.certs, nil
+}
+
+// PrivateKey returns nil for remote signer.
+func (s *remoteSigningMethod) PrivateKey() crypto.PrivateKey {
+	return nil
+}
```

Comment:
Should we through error here?

Reply:
The JWT package signing process asked me to provide a private key which will be passed to our `signingMethod`, however, the remote signer (external plugin) gets the private key from remote key vault store and does not use the private key provided here, so the PrivateKey() return a nil value and it is by design.
```
        token := jwt.NewWithClaims(method, payload)
	token.Header = headers

	// sign and return compact JWS
	compact, err := token.SignedString(method.PrivateKey()) // provide the private key here
	if err != nil {
		return "", nil, err
	}

	// access certificate chain after sign
	certs, err := method.CertificateChain()
	if err != nil {
		return "", nil, err
	}
	return compact, certs, nil
```